packages/project/src/db/tree.ts

```diff
@@ -1,12 +1,113 @@
-import { type Tree, Instance } from "@webstudio-is/react-sdk";
+import { formatAsset } from "@webstudio-is/asset-uploader/server";
+import { type Tree, Instance, Text } from "@webstudio-is/react-sdk";
 import { applyPatches, type Patch } from "immer";
 import {
+  Asset,
   prisma,
   type Prisma,
   type Tree as DbTree,
 } from "@webstudio-is/prisma-client";
 import { utils } from "../index";
-import type { Breakpoint } from "@webstudio-is/css-data";
+import {
+  type Breakpoint,
+  SharedStyleValue,
+  ImageValue,
+} from "@webstudio-is/css-data";
+import { z } from "zod";
+import DataLoader from "dataloader";
+import warnOnce from "warn-once";
+
+const assetsLoader = new DataLoader<string, Asset | undefined>(
+  async (assetIds) => {
+    const assets = await prisma.asset.findMany({
+      where: {
+        id: {
+          in: [...assetIds],
+        },
+      },
+    });
+
+    return assetIds.map((assetId) =>
+      assets.find((asset) => asset.id === assetId)
+    );
+  }
+);
+
+/**
+ * Use zod + DataLoader to load/format assets from the Assets table.
+ */
+const ImageDBAsset = z.object({
+  type: z.literal("asset"),
+  value: z
+    .string()
+    .uuid()
+    .transform(async (assetId) => {
+      const asset = await assetsLoader.load(assetId);
+      if (asset === undefined) {
+        warnOnce(true, `Asset with assetId "${assetId}" not found`);
+        return;
+      }
+
+      return formatAsset(asset);
+    }),
+});
+
+const ImageDbValue = z.object({
+  type: z.literal("image"),
+  value: z
+    .array(ImageDBAsset)
+    // an Asset can be not present in DB, skip it.
+    .transform((assets) => assets.filter((asset) => asset.value !== undefined)),
+});
+
+const StyleDbValue = z.union([SharedStyleValue, ImageDbValue]);
+
+const StyleDb = z.record(z.string(), StyleDbValue);
+
+export const CssDbRule = z.object({
+  style: StyleDb,
+  breakpoint: z.optional(z.string()),
+});
+
+const InstanceDb = z.lazy(() =>
+  z.object({
+    type: z.literal("instance"),
+    id: z.string(),
+    component: z.string(),
+    children: z.array(z.union([InstanceDb, Text])),
+    cssRules: z.array(CssDbRule),
+  })
+) as z.ZodType<Instance>;
+
+/**
+ * In the DB we hold only assetId
+ **/
+const ImageValueDbIn = ImageValue.transform((imageStyle) => ({
+  type: imageStyle.type,
+  value: imageStyle.value.map((value) =>
+    /* Now value.type is always equal to the "asset", but in the future, it will have additional types */
+    value.type === "asset" ? { type: "asset", value: value.value.id } : value
+  ),
+}));
+
+const StyleDbValueIn = z.union([SharedStyleValue, ImageValueDbIn]);
+
+const StyleDbIn = z.record(z.string(), StyleDbValueIn);
+
+export const CssDbInRule = z.object({
+  style: StyleDbIn,
+  breakpoint: z.optional(z.string()),
+});
+
+const InstanceDbIn = z.lazy(() =>
```

Comment:
confusing name, what does it represent?

Reply:
Values that come `In Db`

Reply:
I don't understand, please explain with full sentence

Reply:
Basically here Zod is used to validate and transform.
Just `Db` means - validate and transform Database json to the original (client) type.
`DbIn`means validate and transform `original (client) type` into `Db` type.

Reply:
Added DbIn, DbOut and comments for InstanceDbIn, InstanceDbOut:
`validate/transform client schema into DB data schema.`
`validate/transform DB data schema to the client schema.`